File: 836.rectangle-overlap.cpp
/*
 * @lc app=leetcode id=836 lang=cpp
 *
 * [836] Rectangle Overlap
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool isRectangleOverlap(vector<int>& rec1, vector<int>& rec2) {
    int w1 = rec1[2] - rec1[0];
    int h1 = rec1[3] - rec1[1];
    int w2 = rec2[2] - rec2[0];
    int h2 = rec2[3] - rec2[1];
    if (!w1 || !h1 || !w2 || !h2) {
      return false;
    }
    if (((rec1[0] < rec2[0] && rec2[0] < rec1[2]) ||
         (rec2[0] < rec1[0] && rec1[0] < rec2[2]) ||
         (rec1[0] < rec2[2] && rec2[2] < rec1[2]) ||
         (rec2[0] < rec1[2] && rec1[2] < rec2[2])) &&
        ((rec1[1] < rec2[1] && rec2[1] < rec1[3]) ||
         (rec2[1] < rec1[1] && rec1[1] < rec2[3]) ||
         (rec1[1] < rec2[3] && rec2[3] < rec1[3]) ||
         (rec2[1] < rec1[3] && rec1[3] < rec2[3]))) {
      return true;
    }
    return false;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> rec1, rec2;
  rec1 = {0, 0, 2, 2};
  rec2 = {1, 1, 3, 3};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), true);
  rec1 = {0, 0, 1, 1};
  rec2 = {1, 0, 2, 1};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), false);
  rec1 = {0, 0, 1, 1};
  rec2 = {2, 2, 3, 3};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), false);
  rec1 = {1, 1, 1, 2};
  rec2 = {0, 0, 2, 2};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), false);
  rec1 = {11, 12, 13, 13};
  rec2 = {17, 1, 17, 19};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), false);
  rec1 = {7, 8, 13, 15};
  rec2 = {10, 8, 12, 20};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), true);
  rec1 = {-1, 0, 1, 1};
  rec2 = {0, -1, 0, 1};
  ASSERT_EQ(s.isRectangleOverlap(rec1, rec2), false);
  return 0;
}

File: 836.rectangle-overlap.rs
/*
 * @lc app=leetcode id=836 lang=rust
 *
 * [836] Rectangle Overlap
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn is_rectangle_overlap(rec1: Vec<i32>, rec2: Vec<i32>) -> bool {
        let w1 = rec1[2] - rec1[0];
        let h1 = rec1[3] - rec1[1];
        let w2 = rec2[2] - rec2[0];
        let h2 = rec2[3] - rec2[1];
        if w1 == 0 || h1 == 0 || w2 == 0 || h2 == 0 {
            return false;
        }
        ((rec1[0] < rec2[0] && rec2[0] < rec1[2])
            || (rec2[0] < rec1[0] && rec1[0] < rec2[2])
            || (rec1[0] < rec2[2] && rec2[2] < rec1[2])
            || (rec2[0] < rec1[2] && rec1[2] < rec2[2]))
            && ((rec1[1] < rec2[1] && rec2[1] < rec1[3])
                || (rec2[1] < rec1[1] && rec1[1] < rec2[3])
                || (rec1[1] < rec2[3] && rec2[3] < rec1[3])
                || (rec2[1] < rec1[3] && rec1[3] < rec2[3]))
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::is_rectangle_overlap(
        vec![0, 0, 2, 2],
        vec![1, 1, 3, 3]
    ));
}

#[test]
fn test2() {
    assert!(!Solution::is_rectangle_overlap(
        vec![0, 0, 1, 1],
        vec![1, 0, 2, 1]
    ));
}

#[test]
fn test3() {
    assert!(!Solution::is_rectangle_overlap(
        vec![0, 0, 1, 1],
        vec![2, 2, 3, 3]
    ));
}
